### ISSUE TITLE
round_robin load balancing is not thread-safe (race condition)

### ISSUE THREAD

**Issue by sandlerben** [grpc/grpc#20226]
### What version of gRPC and what language are you using?
Python, 1.21.1

### What operating system (Linux, Windows,...) and version?
MacOS 10.14

### What runtime / compiler are you using (e.g. python version or version of gcc)
Python, 1.21.1

### What did you do?
1. Create a shared channel with a `round_robin` load balancing policy.
2. Make 10 concurrent requests starting at the exact same time.

```
channel = grpc.insecure_channel(URL, [("grpc.lb_policy_name", "round_robin")])

with ThreadPoolExecutor(10) as executor:
   for i in range(10):
       stub = ExampleStub(channel)
       executor.submit(stub.ExampleRequest)
```

### What did you expect to see?

Requests being load balanced and sent to different servers.

### What did you see instead?

All requests go to the same server.

When I add a little bit of a time buffer in the code, the problem goes away (**requests go to different servers**).

```
channel = grpc.insecure_channel(URL, [("grpc.lb_policy_name", "round_robin")])

with ThreadPoolExecutor(10) as executor:
   for i in range(10):
       stub = ExampleStub(channel)
       executor.submit(stub.ExampleRequest)
       time.sleep(1)
```

My theory is that the underlying load balancing code is not thread safe (e.g. the round robin isn't updated by the first thread making the request in time for the second thread to see it).

**Comment by dsethi** [comment]
Not sure if this is the right place for this question:

Is batching implemented in python? My use case is sending a large number of small protobufs over rpc -- I need to batch them for throughput. Sample python client code -- which does not work (I.e. no speedup by changing any of the option values) --

channel = grpc.insecure_channel('localhost:2020', options=[('grpc.max_concurrent_streams', 1),
                                                           ('grpc.max_send_message_length', 1 * 1024 * 1024),
                                                           ('grpc.max_receive_message_length', 1 * 1024 * 1024),
                                                           ('grpc.http2.max_frame_size', 16777215),
                                                           ('grpc.http2.write_buffer_size', 10 * 1024 * 1024),
                                                           ('WriteFlag.buffer_hint', True)
Am I missing something?

**Comment by AspirinSJL** [comment]
@sandlerben The round_robin implementation is thread safe. The reason that you are seeing all the requests going to the same server is that connection to server needs time. In details, when you start connecting to 10 severs, the connection attempts are in parallel. When the first server is connected, we will have a ready server pool of size 1. So all your requests go to that server. When more servers are connected, the pool will include them one by one. If you add some time buffer, I guess all the 10 servers are connected and are in the pool, so your requests will be load balanced as expected. 

Please re-open if you have further questions.

**Comment by AspirinSJL** [comment]
@dsethi Please open a separate issue for proper tracking. Thanks!

**Comment by sandlerben** [comment]
@AspirinSJL is there any way to have the connections open when the application starts or just before requests are made?

**Comment by AspirinSJL** [comment]
Nope, we don't have any way to make sure all the servers are connected before requests are made. In principle, only the connectivity of the whole channel is visible to users.The load balancing part should be transparent.

**Comment by sandlerben** [comment]
That makes sense. Do you think there any good ways to support this use-case, though? I imagine it is shared by other users (it's a pain shared by others at my company).

**Comment by AspirinSJL** [comment]
What's the reason to have this feature? I assume you want perfect balancing among your backend servers. If that's the case, will you allow RPCs to be issued if any previously connected backend fails?

**Comment by AspirinSJL** [comment]
cc @markdroth

**Comment by markdroth** [comment]
To say this another way, I think it would be very bad for reliability to have the client not send any RPCs until all of the backends are up, beacuse that would mean that a single backend being down would cause all RPCs to fail.  I don't think this feature request is a good idea.

**Comment by sandlerben** [comment]
I agree with you, I wouldn't want to wait for "all" back-ends (how can the client even know the number of back-ends). I guess there's really no way around it because all the requests should be sent the moment any connection is made. I think the best solution, if any, is a way to have the `grpc` client exposes functionality to try to make connections. That can be called explicitly before requests are made (e.g. when application starts up).

**Comment by markdroth** [comment]
We already have an API for asking a channel to connect before you send an RPC on it.  It's part of the channel state API.  @lidizheng or @gnossen can probably point you at the right API to use for that in Python.

**Comment by sandlerben** [comment]
Fantastic that's exactly what I need. Looking forward to turning that on.

**Comment by lidizheng** [comment]
@sandlerben I think the API you are looking for is [`grpc.channel_ready_future`](https://grpc.github.io/grpc/python/grpc.html#channel-ready-future).

**Comment by sandlerben** [comment]
Perfect, that worked!
